packages/itwin/measure-tools/src/measurements/PerpendicularDistanceMeasurement.ts

```diff
@@ -0,0 +1,231 @@
+    const heightPoints = this.getHeightPoints([customStartPoint ?? this.startPointRef, Point3d.createFrom(point)]);
+    let newEndPoint = point;
+    if (this._measurementType === PerpendicularMeasurementType.Height) {
+      newEndPoint = heightPoints[1];
+    }
+    super.setEndPoint(newEndPoint);
+    this.updateSecondaryLine(heightPoints);
+  }
+
+  private updateSecondaryLine = (heightPoints: Point3d[]) => {
+    switch (this._measurementType) {
+      case PerpendicularMeasurementType.Width:
+        this._secondaryLine = [heightPoints[0], heightPoints[1]];
+        break;
+      case PerpendicularMeasurementType.Height:
+        this._secondaryLine = [heightPoints[1], heightPoints[2]];
+        break;
+    }
+  };
+
+  /**
+   * Returns the points for the base and perpendicular lines of a right triangle formed from the hypotenuse.
+   */
+  private getHeightPoints(hypotenusePoints: Point3d[]): Point3d[] {
+    const heightPoints: Point3d[] = [];
+    heightPoints.push(hypotenusePoints[0].clone());
+    if (hypotenusePoints[0].z > hypotenusePoints[1].z) {
+      heightPoints.push(new Point3d(hypotenusePoints[0].x, hypotenusePoints[0].y, hypotenusePoints[1].z));
+    } else {
+      heightPoints.push(new Point3d(hypotenusePoints[1].x, hypotenusePoints[1].y, hypotenusePoints[0].z));
+    }
+    heightPoints.push(hypotenusePoints[1].clone());
+
+    return heightPoints;
+  }
+
+  public static override create(start: Point3d, end: Point3d, viewType?: string) {
+    // Don't ned to serialize the points, will just work as is
+    const measurement = new PerpendicularDistanceMeasurement({ startPoint: start, endPoint: end });
+    if (viewType) {
+      measurement.viewTarget.include(viewType);
+    }
+
+    return measurement;
+  }
+
+  public override decorate(context: DecorateContext): void {
+    super.decorate(context);
+
+    if (this._secondaryLine && this._secondaryLine.length > 0) {
+      const secondaryLine = context.createGraphicBuilder(GraphicType.WorldOverlay, undefined, this.getSnapId());
+      secondaryLine.setSymbology(ColorDef.white, ColorDef.black, 1, LinePixels.Code5);
+      secondaryLine.addLineString(this._secondaryLine);
+      context.addDecorationFromBuilder(secondaryLine);
+    }
+  }
+
+  protected override async getDataForMeasurementWidgetInternal(): Promise<MeasurementWidgetData> {
+    const distanceData = await super.getDataForMeasurementWidgetInternal();
+
+    const lengthSpec = await IModelApp.quantityFormatter.getFormatterSpecByQuantityType(QuantityType.LengthEngineering);
+    const distance = this.worldScale * this.startPointRef.distance(this.endPointRef);
+    const fDistance = IModelApp.quantityFormatter.formatQuantity(distance, lengthSpec);
+
+    const title =
+      this._measurementType === PerpendicularMeasurementType.Height
+        ? MeasureTools.localization.getLocalizedString("MeasureTools:tools.MeasureHeight.toolTitle").replace("{0}", fDistance)
```

Comment:
@saaaaaally iTwin localization allows passing arguments when getting the localized string, e.g. instead of `{0}` you can have it as `{{value}}` and pass in a json object with `{ value: fDistance }`. that's more robust than doing a string replace here.

Reply:
Done

Reply:
@saaaaaally You don't need to do the string replace still, `getLocalizedString`'s second argument is a [TranslationOptions](https://www.itwinjs.org/reference/core-common/localization/translationoptions/), and the localization library will replace `{{value}}` automatically when you pass in `{value: fDistance}`.

https://www.itwinjs.org/learning/frontend/localization/

Reply:
Fixed